core/src/main/java/hudson/util/ProcessTree.java

```diff
@@ -512,10 +516,38 @@ public void kill() throws InterruptedException {
             }
             
             LOGGER.log(FINER, "Killing {0}", getPid());
+            // Firstly try to kill it gracefully, then do a forcekill if it does not help (algorithm is described in JENKINS-17116)
+            killSoftly();
             p.kill();
```

Comment:
Would it be possible to determine whether the process got killed softly (e.g. by boolean return)? If yes, maybe it makes not sense to invoke other operations

Reply:
Inside killSoftly we use isRunning to find out whether the process is around still. If not, we could skip the kill. But kill won't have any effect any way and I'd like to be cautious here and just add killSoftly as an extra step in addition to the old kill code.